r/synthetics_canary - new resource

DrFaust92 opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #11145

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource_aws_synthetics_canary: add new service
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSSyntheticsCanary_'
--- PASS: TestAccAWSSyntheticsCanary_tags (349.62s)
--- PASS: TestAccAWSSyntheticsCanary_disappears (146.87s)
--- PASS: TestAccAWSSyntheticsCanary_runConfig (213.57s)
--- PASS: TestAccAWSSyntheticsCanary_basic (86.61s)
--- PASS: TestAccAWSSyntheticsCanary_runtimeVersion (208.80s)
--- PASS: TestAccAWSSyntheticsCanary_runConfigTracing (208.81s)
```

DrFaust92 commented:
basic use case test now passing

need to add the following:
* more tests (VPC and other configs)
* add logic to get code from s3 (only arguments are added)
* wait for creation (maybe with flag)

DrFaust92 commented:
When a canary is deleted, the Lambda that is created by it is not deleted...
and when that is coupled with a VPC config there is an endless delete loop for VPC resources as the lambdas ENIs are also not deleted

 a work around is to delete the function as well but it requires calling the lambda from the canary resource for a resource thats not managed by terraform (albeit it is a byproduct)

@bflad/ @ewbankkit , thoughts?

bflad commented:
@DrFaust92 sounds less than ideal. I think first things first we should split this PR into the service client and tagging parts (which we can merge immediately) versus the problematic resource implementation (which I'm not sure when we'll have more time to look into this issue).

DrFaust92 commented:
Will split these

DrFaust92 commented:
opened for only service and tag implementation #13186

tdmalone commented:
> for a resource thats not managed by terraform (albeit it is a byproduct)

@DrFaust92 This sounds similar - although admittedly, more complex - to eg. a CloudWatch log group being created the first time a Lambda is run and not being deleted when the Lambda is deleted. The general workaround for that with Terraform is to also define the log group, and ensure it's created before the function runs for the first time; the function then uses that log group and if you delete everything from Terraform, everything is deleted.

I haven't played around with Synthetics in anger yet so I'm not sure if a similar workflow would work (particularly given you probably wouldn't want to manage the entire Lambda from within Terraform... unless the code can be easily referenced from an S3 object? 🤔) but perhaps that workflow ^^ can offer inspiration towards a solution here.

immo-huneke-zuhlke commented:
I've used the [temporary provider based on this pull request](https://github.com/MaksymBilenko/terraform-provider-aws-synthetics) with good results. I'd be keen to see this merged ASAP because the temporary provider will not pass the security review before production release, being considered SOUP (software of unknown provenance). Any idea when it might happen?

DrFaust92 commented:
hey @immo-huneke-zuhlke, i wasn't looking at this lately but at the time i was having issues with canary resource deletion as there are resource that are implicitly created by canary and create annoying dependencies. if you are satisfied with a fork of this branch im guessing some changes were made to it.

i'll be happy receive a review of what changes are required to make this work or merge change from the fork.

immo-huneke-zuhlke commented:
Hi Ilia,

Thanks for getting back to me so quickly.

The truth is that I don’t yet know if there are any resources left behind that have to be cleaned up manually – maybe there are some that I have not noticed.

I create the Lambda execution role and the associated policy explicitly, so those get deleted when the canary is destroyed. However, the role has permission to create Cloudwatch log groups and log streams, so any that it creates on the fly will not be deleted automatically. I think that’s ok, because logs should be accessible even after their originator has been deleted.

Similarly the outputs generated by the synthetic (HAR files, screenshots and stdout/stderr) are placed in the S3 bucket created for the synthetic canary by the Terraform script – but since this bucket is deleted when the canary is destroyed, the outputs disappear with it.

I think the documentation should warn users about these behaviours, but they are quite acceptable.

Best regards,
Immo

From: Ilia Lazebnik <notifications@github.com>
Reply to: terraform-providers/terraform-provider-aws <reply@reply.github.com>
Date: Thursday, 16 July 2020 at 11:33
To: terraform-providers/terraform-provider-aws <terraform-provider-aws@noreply.github.com>
Cc: Immo Huneke <immo.huneke@zuhlke.com>, Mention <mention@noreply.github.com>
Subject: Re: [terraform-providers/terraform-provider-aws] [WIP]r/synthetics_canary - new resource (#13140)


hey @immo-huneke-zuhlke<https://github.com/immo-huneke-zuhlke>, i wasn't looking at this lately but at the time i was having issues with canary resource deletion as there are resource that are implicitly created by canary and create annoying dependencies. if you are satisfied with a fork of this branch im guessing some changes were made to it.

i'll be happy receive a review of what changes are required to make this work or merge change from the fork.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/terraform-providers/terraform-provider-aws/pull/13140#issuecomment-659325517>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACHOGYF2U22VFUQMA2MLENLR33JRHANCNFSM4MYIXA5Q>.

ewbankkit commented:
@DrFaust92 I have been testing with what we have in this PR 👏 and it seems like once the Synthetics Canary has been created I can manually delete the Lambda that was implicitly created in the console. A `terraform destroy` of the in-VPC acceptance test configuration then completes cleanly.
The answer may then be to add an acceptance test step that deletes the lambda (there is a standard naming pattern [documented](https://docs.aws.amazon.com/AmazonSynthetics/latest/APIReference/API_DeleteCanary.html)).
There is still the problem that in a pure IaC world that lambda can't be deleted via this provider. The user will have to shell out to the AWS CLI or something.
As Amazon do, we could add the list of implicitly created resources to the documentation and even return the Lambda function and layer ARNs as `Computed attributes`.

DrFaust92 commented:
Added some docs regarding implicit resources + some added tests (disappears)
added support for setting memory for canary lambda.

still experimenting with lambda deletion but all else seems to be working after a few tweaks.

MaksymBilenko commented:
I think that canary start option is missing here.

DrFaust92 commented:
@MaksymBilenko, not sure i want to mix canary execution and canary start in a single resource. id split it it of to a different data source maybe, like lambda execution.

brntsllvn commented:
Looking at the PR, this looks mostly done...? I'm happy to beta test or contribute if that moves things forward. Let me know how I can assist. Thanks for your effort here.

DrFaust92 commented:
rebased

DrFaust92 commented:
```
--- PASS: TestAccAWSSyntheticsCanary_tags (349.62s)
--- PASS: TestAccAWSSyntheticsCanary_disappears (146.87s)
--- PASS: TestAccAWSSyntheticsCanary_runConfig (213.57s)
--- PASS: TestAccAWSSyntheticsCanary_basic (86.61s)
```

olysyuk commented:
I've used code from PR and it works ok for canary creation. But once canary has been run it can no longer update it

```
Error: error waiting for Synthetics Canary (test-synthetics) updating: unexpected state 'STOPPED', wanted target 'READY'. last error: %!s(<nil>)
```

DrFaust92 commented:
reminder to self, for updating running canaries will have to stop the canary first.

MaksymBilenko commented:
> reminder to self, for updating running canaries will have to stop the canary first.

What about this kind of approach
https://github.com/MaksymBilenko/terraform-provider-synthetics/blob/master/synthetics/resource_aws_synthetics_canary.go#L395

DrFaust92 commented:
@MaksymBilenko this is similar to what i had in mind. im reworking the PR now to add more tests and make the VPC tests finish (needs to remove implicit lambda layer and lambda) and then ill work on this

MaksymBilenko commented:
@DrFaust92 For proper VPC configuration update in-VPC/Non-VPC/VPC migration - I had issues and fixed by this:
https://github.com/MaksymBilenko/terraform-provider-synthetics/blob/master/synthetics/resource_aws_synthetics_canary.go#L134

DrFaust92 commented:
I'm not having issues with update as much as when deleting the canary upstream resources like subnet cannot be sweeped by terraform as there are lingering resources creating ENIs like the underlying lambda.

DrFaust92 commented:
Handle the following:

- [] allow starting/stopping Canary
- [] Update/Delete Running Canary

DrFaust92 commented:
added starting/stopping canary and fixed waiters. lint fail unrelated.

immo-huneke-zuhlke commented:
We have encountered a problem when using Dr Maksym Bilenko's third-party AWS provider that incorporates this pull request. When the source code of the synthetic canary has changed, Terraform doesn't detect this and deploy an updated version.

The aws_lambda_function resource supports a property source_code_hash, which is an efficient mechanism for triggering an update. Could this be added to the synthetics_canary resource too?

My workaround has been to use two null_resource resources: one to detect changes in the source code and delete the local zip file, and a second one to delete changes in the zip file and trigger a redeployment of the synthetics_canary. Very clumsy, but if anyone needs to know how it's done please ask.

DrFaust92 commented:
Hey @immo-huneke-zuhlke, ill try to test this in my branch (and possibly add a test for this) and post here with results. many thanks for bringing this up.

DrFaust92 commented:
added test for local code modification, changing the code on this branch updates the canary.

DrFaust92 commented:
Added sweeper (also deletes orphan lambda functions and layers)

DrFaust92 commented:
added support for setting `runtime_version` and `active_tracing` in the `run_config` block.

ewbankkit commented on website/docs/r/synthetics_canary.html.markdown:
```diff
@@ -0,0 +1,98 @@
+* `runtime_version` - (Required) Specifies the runtime version to use for the canary. Currently, the only valid values are `syn-nodejs-2.0`, `syn-nodejs-2.0-beta`, and `syn-1.0`. For more information about runtime versions, see [Canary Runtime Versions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html).
+* `artifact_s3_location` - (Required) The location on Amazon S3 where Synthetics stores artifacts from the test runs of this canary.
+* `schedule` -  (Required) Information about how often the canary is to run and when these test runs are to stop. See [Schedule](#schedule) below.
+* `handler` - (Required) The domain description.
```
```suggestion
* `handler` - (Required) The entry point to use for the source code when running the canary. This value must end with the string `.handler` .
```

ewbankkit commented on website/docs/r/synthetics_canary.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `name` - (Required) The name for this canary.
+* `runtime_version` - (Required) Specifies the runtime version to use for the canary. Currently, the only valid values are `syn-nodejs-2.0`, `syn-nodejs-2.0-beta`, and `syn-1.0`. For more information about runtime versions, see [Canary Runtime Versions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html).
+* `artifact_s3_location` - (Required) The location on Amazon S3 where Synthetics stores artifacts from the test runs of this canary.
```
```suggestion
* `artifact_s3_location` - (Required) The location in Amazon S3 where Synthetics stores artifacts from the test runs of this canary.
```

ewbankkit commented on aws/resource_aws_synthetics_canary.go:
```diff
@@ -0,0 +1,649 @@
+					return strings.TrimPrefix(new, "s3://") == old
+				},
+			},
+			"handler": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"s3_bucket": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+				RequiredWith:  []string{"s3_key"},
+			},
+			"s3_key": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+				RequiredWith:  []string{"s3_bucket"},
+			},
+			"s3_version": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+			},
+			"zip_file": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"s3_bucket", "s3_key", "s3_version"},
+			},
```
Should live in a `code` configuration block to match the [AWS API](https://docs.aws.amazon.com/AmazonSynthetics/latest/APIReference/API_CreateCanary.html#API_CreateCanary_RequestSyntax).

DrFaust92 replied:
Cant do that, getting diffs as these arguments are not returned by API. this is handled similarly to how to lambda functions. there is a `FunctionCode` object in the API but its in the root of the resource schema.

ewbankkit commented on aws/internal/service/synthetics/waiter/status.go:
```diff
@@ -0,0 +1,29 @@
+
+func CanaryStatus(conn *synthetics.Synthetics, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		input := &synthetics.GetCanaryInput{
+			Name: aws.String(name),
+		}
+
+		output, err := conn.GetCanary(input)
```
Add `CabaryByName` to a `finder` package and use here, resource Read and acceptance tests.

robertstettner commented on aws/resource_aws_synthetics_canary.go:
```diff
@@ -0,0 +1,649 @@
+	resp, err := conn.GetCanary(input)
+	if err != nil {
+		if isAWSErr(err, synthetics.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] CodeCommit Repository (%s) not found, removing from state", d.Id())
```
```suggestion
			log.Printf("[WARN] Synthetics Canary (%s) not found, removing from state", d.Id())
```

ewbankkit commented:
FYI, [Synthetics now supports the use of environment variables with canaries](https://aws.amazon.com/about-aws/whats-new/2020/11/amazon-cloudwatch-synthetics-supports-environment-variables/).

DrFaust92 commented:
> FYI, [Synthetics now supports the use of environment variables with canaries](https://aws.amazon.com/about-aws/whats-new/2020/11/amazon-cloudwatch-synthetics-supports-environment-variables/).
> 
> The required AWS SDK update has been merged: #16234.

Cannot be read after apply so ill hold of on this for now. otherwise rebased and tested

hamza-tumturk commented:
Is there any plan for releasing this any time soon?

ewbankkit commented:
Addressing this issue is on the quarter's [Roadmap](https://github.com/hashicorp/terraform-provider-aws/blob/main/ROADMAP.md#cloudwatch-synthetics).

YakDriver commented on website/docs/r/synthetics_canary.html.markdown:
```diff
@@ -0,0 +1,101 @@
+* `vpc_config` - (Optional) Configuration block. Detailed below.
+* `zip_file` - (Optional) ZIP file that contains the script, if you input your canary script directly into the canary instead of referring to an S3 location. It can be up to 5 MB. **Conflicts with `s3_bucket`, `s3_key`, and `s3_version`.**
+
+### schedule
```
Since the modern documentation has links on the right side of the page to all the sections, internal links to the subsections aren't needed.

YakDriver reviewed (approved):
Looks great! 🎉 

GovCloud:
```
--- PASS: TestAccAWSSyntheticsCanary_disappears (36.45s)
--- PASS: TestAccAWSSyntheticsCanary_s3 (58.41s)
--- PASS: TestAccAWSSyntheticsCanary_runConfigTracing (86.62s)
--- PASS: TestAccAWSSyntheticsCanary_basic (92.23s)
--- PASS: TestAccAWSSyntheticsCanary_runtimeVersion (92.29s)
--- PASS: TestAccAWSSyntheticsCanary_tags (97.04s)
--- PASS: TestAccAWSSyntheticsCanary_runConfig (98.37s)
--- PASS: TestAccAWSSyntheticsCanary_startCanary (101.86s)
--- PASS: TestAccAWSSyntheticsCanary_vpc (477.29s)
```

`us-west-2`:
```




--- PASS: TestAccAWSSyntheticsCanary_basic (86.26s)
--- PASS: TestAccAWSSyntheticsCanary_disappears (40.82s)
--- PASS: TestAccAWSSyntheticsCanary_runConfig (80.27s)
--- PASS: TestAccAWSSyntheticsCanary_runConfigTracing (78.04s)
--- PASS: TestAccAWSSyntheticsCanary_runConfigTracing (80.96s)
--- PASS: TestAccAWSSyntheticsCanary_runtimeVersion (75.07s)
--- PASS: TestAccAWSSyntheticsCanary_s3 (45.80s)
--- PASS: TestAccAWSSyntheticsCanary_startCanary (92.94s)
--- PASS: TestAccAWSSyntheticsCanary_tags (76.34s)
--- PASS: TestAccAWSSyntheticsCanary_vpc (655.78s)
```
